### ISSUE TITLE
Web - Assign task - New task replies not showing in the main chat between task creator and assignee

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#24532]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Make sure you have at least 3 accounts created.
2. Log into two of the accounts in Mac Chrome (one may have to be in an incognito window). You'll be "User A".
3. Open a chat with "User B".
4. Click on the big, green, circular button with a + in it (FAB) and choose "Assign task".
5. Add a title, select "User B" as the assignee, select "User C" as the "Share somewhere" option, and then click on the green "Confirm task" option.
6. Send a couple of messages in the newly-created task thread while, as "User B", observing the task created in your main chat with "User A".

## Expected Result:
The number of replies in the task thread is shown and updates accordingly in "User B's" main chat with "User A".

## Actual Result:
The replies in the task thread only show after the user navigates to a different chat and back, and it doesn't update accordingly.

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** v1.3.53-1

**Reproducible in staging?:** Y

**Reproducible in production?:** Y

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/115492554/8f08ac38-9ee6-4faa-95e0-3ba73f9d5dad

https://github.com/Expensify/App/assets/115492554/213b9016-7679-4c13-9a9c-53fc2c8c1d7d

https://github.com/Expensify/App/assets/115492554/61e606ab-74bf-4424-8c67-2185b05fc210

**Expensify/Expensify Issue URL:**

**Issue reported by:** @Vicktor-nyagudi

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1691146112430489

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01628e3a90daf05d44</li>
        <li>Upwork Job ID: 1696762480991907840</li>
        <li>Last Price Increase: 2023-08-30</li>
    </ul>
</details>

**Comment by ShogunFire** [comment]
@s-alves10 can you confirm here, you think this is missing pusher events ? This is what I receive:
![image](https://github.com/Expensify/App/assets/19537677/7a70b42c-7044-4e6b-ba1d-4c6011953992)
It''s not enough to do the link ?

**Comment by s-alves10** [comment]
@ShogunFire 

Yeah. That's not enough.
For task report, we have two parent report: shareDestination and assignee(Can be 1 if they are same). When we send messages in the task report, we need the pusher events that notify these parent report actions. It looks like you get the pusher events for task report itself.

**Comment by luacmartins** [comment]
This seems to be internal IF we can reproduce and it requires additional pusher updates

**Comment by JmillsExpensify** [comment]
Hmm, yeah interesting. Triaging for internal.

**Comment by JmillsExpensify** [comment]
Still need someone to pick this up.

**Comment by joelbettner** [comment]
Sorry, I have too much on my plate to pick this up right now. I'm buried in dailies and still getting hourlies assigned to me on a regular basis. I'm going to un-assign myself.

**Comment by JmillsExpensify** [comment]
Still waiting for someone internal.

**Comment by JmillsExpensify** [comment]
Still not an internal priority, waiting for an internal engineer to pick this up.

**Comment by JmillsExpensify** [comment]
Same, we need an internal engineer to pick up this issue.

**Comment by JmillsExpensify** [comment]
Still waiting on internal engineers. Moving to monthy

**Comment by lakchote** [comment]
Taking it and putting it as a weekly.

**Comment by JmillsExpensify** [comment]
Same as above

**Comment by JmillsExpensify** [comment]
Same same

**Comment by lakchote** [comment]
Sorry for the lack of updates, but I'm really putting the focus on budgeting right now.

**Comment by JmillsExpensify** [comment]
No worries, that's understandable

**Comment by lakchote** [comment]
Gave it a look today, didn't figure it out yet how to send the correct Pusher event to update correctly the number of task replies.
I've tried sending a pusher event with the correct `childVisibleActionCount` (number of task replies), along with notifying the correct participant (the task assignee) so he receives the event, he receives the event but the number of task replies don't increase. Which makes me think it must be a problem of using the right report ID/reportActionID.

Will update whenever I have something new.

**Comment by lakchote** [comment]
I'll have to go on parental leave, but I found a lead for the person that will pick this up.

Basically if we go on a report page like: `https://dev.new.expensify.com:8082/r/5315000601989737`
The report ID in the route is 5315000601989737.

If we assign a task, we'll need to reuse the reportActionID for that.
Example:
```
2023-12-21 15:49:28.384  5315000601989737  3          ADDCOMMENT  {"html":"assigned to y6@expensify.com","lastModified":"2023-12-21 15:49:28.384","taskReportID":8070629696725126}  4790341435699635022
```
In that case we'll reuse the reportActionID `4790341435699635022`.

Then, with these two elements we're able to correctly update the task replies in the assignee side.
In `sendChildDetailsToParentAction()`, `$parentParticipants` will need to be updated and include the assignee account ID on top of the task creator's account ID. In our example, that'd be `[3,6]`. 3 is the task creator and 6 is the assignee.

So for example, if the task creator is account 
For `$parentReportID` use the report ID in the route (5315000601989737).
For `$parentReportActionID` use the reportActionID (4790341435699635022).

That'd give something like this (hardcoded so you get precisely what I'm illustrating):
```PHP
        $parentParticipants = [3, 6];
        $onyxData = [[
            'onyxMethod' => Onyx::METHOD_MERGE,
            'key' => OnyxKeys::COLLECTION_REPORT_ACTIONS.'5315000601989737',
            'value' => [
                '4790341435699635022' => [
                    'childCommenterCount' => $response['childCommenterCount'] ?? null,
                    'childLastVisibleActionCreated' => $response['childLastVisibleActionCreated'] ?? null,
                    'childLastActorAccountID' => $response['childLastActorAccountID'] ?? null,
                    'childOldestFourAccountIDs' => $response['childOldestFourAccountIDs'] ?? null,
                    'childVisibleActionCount' => $response['childVisibleActionCount'] ?? null,
                    'childMoneyRequestCount' => $response['childMoneyRequestCount'] ?? null,
                    'childLastMoneyRequestComment' => $response['childLastMoneyRequestComment'] ?? null,
                    'childLastReceiptTransactionIDs' => $response['childLastReceiptTransactionIDs'] ?? null,
                    'childReportNotificationPreference' => $response['childReportNotificationPreference'] ?? null,
                    'childRecentReceiptTransactionIDs' => $response['childRecentReceiptTransactionIDs'] ?? null,
                    'childType' => $response['childType'] ?? null,
                ],
            ],
        ]];
```

If you do it, you'll notice the number of replies correctly increase on the assignee's side.
However, it does not increase for the task creator, so you'll need to figure that out.

Hopefully that gives you a lead that is going to save you some time.

**Comment by JmillsExpensify** [comment]
@ntdiary are you the lead who is picking this one up?

**Comment by JmillsExpensify** [comment]
Ok thank you!

**Comment by JmillsExpensify** [comment]
@lakchote Are you going to pick this up? If not, I'm inclined to close it since tasks are not a roadmap priority at the moment.

**Comment by lakchote** [comment]
@JmillsExpensify I'm going to focus on waves to speed up the reunification. I can pick it up later if tasks become a priority?

**Comment by lakchote** [comment]
I'm putting this up as a monthly since I found a lead and I'd like to finish it.

**Comment by lakchote** [comment]
Had 2 days where I could focus on it for a few hours, but didn't find the culprit yet.

Asked @luacmartins for help as he has more experience in Onyx than me.

cc @chiragsalian too, since you're the one who created the `sendChildDetailsToParentAction()`, maybe you'll notice something right from the start?

**Comment by chiragsalian** [comment]
I haven't investigated but in Action Performed -> Step 6, did you check,
1. Does the onyx data for the message contains userA as one of the senders?

My guess is userA is not one of the sender when new messages are added by userB. This is most likely because userA either,
1. Does not have the report shared with them, Or
2. Their notification setting for the report is different than what userB and userC has it.

Hence when userA clicks to another report and clicks back, i imagine the openReport auth command applies the correct setting. Basically if you check userA sharedReport and notification setting before and after the openReport command when they click back into the task report.

I haven't deep dived but let me know if that helps ✌️

**Comment by lakchote** [comment]
Thanks Chirag for your input.

Actually the problem is both for user B, and user C.

The notification setting is set to `always` for all three of them.

The participant account IDs in `sendChildDetailsToParentAction()` concerns both User A and User C, yet it's not updated for user C.

Regarding the report permissions in the `sharedReports` table, User A and User B have `read, write`.

So I'm still looking to see where the problem is.

**Comment by chiragsalian** [comment]
Cool sounds good, if you are stuck let me know and ill test the flow and debug as well.

**Comment by lakchote** [comment]
> Cool sounds good, if you are stuck let me know and ill test the flow and debug as well.

Thank you @chiragsalian for your proposal! I'm going to be ooo tomorrow for the day, and on Monday I'll need to focus on writing the wave 8 Simplified Collect Xero detailed implementation to get this started.

_If_ you have the bandwidth I would really appreciate it if you could take a look at this. 
I've tried to spot differences between different reports or if any participant account ID is missing, without luck. 
Having this issue solved and knowing what was the culprit would be great, I'd definitely subscribe to it! Let me know if that's okay for you, otherwise I'll pick it up again whenever I can :)

**Comment by chiragsalian** [comment]
Sorry bud, my plate is a bit full too and I'm avoiding picking up VIP issues over wave issues unless the VIP issue is pretty important. If the priority here increases or if my plate gets a bit empty i can jump back here but for now i wont be picking this up.

**Comment by lakchote** [comment]
> I've gave it another look today.
> 
> 1. I've created a task report with report action ID `4784439244454437068` and report ID `2392293761951644`.
> 2. I've checked which onyx update we're sending when we're adding a comment, it's for the key: `reportActions_2392293761951644` with value `4784439244454437068`.
>    The thread replies number is correctly sent and incremented.
> 3. So in theory, it should update the thread replies correctly, but it doesn't.
>    The `childVisibleActionCount` keeps the old value until we go back to the parent report of the task report.
>    In OpenReport response, we get (using the same Onyx merge update `reportActions_2392293761951644` with value `4784439244454437068` combo), the correct `childVisibleActionCount` and we get the updated number of thread replies. So my open question: why doesn't it update at first the thread replies count?
> 
> I've DMed @danieldoglas who seems well-versed with Onyx updates, to understand what I'm missing in order to move this forward.

So, it seems like after discussing with @danieldoglas that my reasoning is correct and thread replies should increment.

He told me to look in the `onyxUpdates` table and see what is being sent in the report channel.
Well, new thread replies number in `childVisibleActionCount` is correctly sent.

I've DMed @s-alves10 to ask about those missing pusher updates he had in mind. Maybe I've missed something.

**Comment by JmillsExpensify** [comment]
Nice, looks great! So I believe that the main next steps would be payments. This issue was created before we stopped paying for bug bounties, so I believe that @Victor-Nyagudi would be owed $250 for the bug report. Given that no PR was reviewed, I don't believe that any C+ payment would be due.

**Comment by JmillsExpensify** [comment]
All paid out. Thanks!
